### ISSUE TITLE
haveno: add package

### ISSUE THREAD

**Issue by milahu** [milahu/nixpkgs#68]
fix https://github.com/NixOS/nixpkgs/issues/293842

based on https://github.com/NixOS/nixpkgs/pull/311314
based on https://github.com/tadfisher/gradle2nix/pull/62

draft in [haveno/default.nix](https://github.com/milahu/nur-packages/raw/master/pkgs/applications/blockchains/haveno/default.nix)

```
nix-shell -p nur.repos.milahu.haveno
```

### haveno is not working

the desktop app is working, but it fails to connect to seednodes
i dont know whether this is a problem with the client or a problem with the haveno network

the gui statusbar says: no seed nodes available

the seednode seems offline:
bugc27z7lqjgpwmhbuu3kuwoq2bhailj573r32jm5ydwrcqrwjtblnid.onion:1002

```
$ torsocks nc bugc27z7lqjgpwmhbuu3kuwoq2bhailj573r32jm5ydwrcqrwjtblnid.onion 1002
1720959904 ERROR torsocks[402405]: Host unreachable (in socks5_recv_connect_reply() at socks5.c:539)

$ curl --proxy socks5h://user$RANDOM@127.0.0.1:9050 http://bugc27z7lqjgpwmhbuu3kuwoq2bhailj573r32jm5ydwrcqrwjtblnid.onion:1002
curl: (97) Can't complete SOCKS5 connection to bugc27z7lqjgpwmhbuu3kuwoq2bhailj573r32jm5ydwrcqrwjtblnid.onion. (4)
```

the main version of haveno is broken by design
because the authors want to separate development from deployment
so im using the [haveno-reto](https://github.com/retoaccess1/haveno-reto) fork
which has seednodes and different keys
but otherwise its exactly the same code

### is haveno a scam

it is possible that haveno is a scam, a waste of time

https://old.reddit.com/r/Monero/comments/1cqf02p/haveno_mainnet_documentation_released/

> its already running on mainnet however not usable because there's no mods.

> how do you propose to run a mainnet network from the official repository without being exposed to significant legal risk?

https://old.reddit.com/r/Monero/comments/1e60nc8/haveno_instances_experience/

> Haveno Reto is the only running Haveno network that everybody knows about

https://github.com/haveno-dex/haveno/issues/1142

> im [building](https://github.com/milahu/nur-packages/raw/master/pkgs/applications/blockchains/haveno/default.nix) haveno from source, but i guess it requires more patching
> maybe there are some more libs or bins hidden in jars
> which dont run on nixos by default and cause silent errors
> but i rather invest my time in [basicswap](https://github.com/basicswap/basicswap)

> having a list of N seednodes is not "decentralized"
> haveno is a pipe dream similar to [openbazaar](https://en.wikipedia.org/wiki/OpenBazaar)

[haveno/readme.txt](https://github.com/milahu/nur-packages/raw/master/pkgs/applications/blockchains/haveno/readme.txt)

### future work

i already spent too much time on this...

if someone wants to get their hands dirty on this
i would recommend to first try a binary release of [haveno-reto](https://github.com/retoaccess1/haveno-reto)
maybe with `steam-run`, maybe with [buildFHSEnv](https://ryantm.github.io/nixpkgs/builders/special/fhs-environments/)
maybe on a different linux distro like [arch linux](https://aur.archlinux.org/packages/haveno-reto)
to verify that `haveno-reto` is actually working

then i would modify [gradle2nix](https://github.com/tadfisher/gradle2nix/pull/62) so that my [mkOverride](https://github.com/milahu/nur-packages/raw/master/pkgs/applications/blockchains/haveno/default.nix) function is applied to all jar files
maybe there are still some libs or bins that need to be patched
otherwise `haveno` silently fails... but that is an optimistic view 

ping @Kreyren @linsui @emmanuelrosa @chayleaf @surfaceflinger @TanvirOnGH @shortwavesurfer2009 @nahuhh

**Comment by Kreyren** [comment]
Why don't you submit the code as merge request in nixpkgs? Makes it easier to review

**Comment by linsui** [comment]
Because they are banned.

**Comment by TanvirOnGH** [comment]
> Because they are banned.

Reason?

**Comment by linsui** [comment]
https://github.com/NixOS/moderation/commit/836964980e170fe6c856eb0296c27b8337309be8

**Comment by milahu** [comment]
"they" xD
i was banned because i give a shit about "modern feelings", i only care about code

i still think its interesting whether `haveno-reto` is actually working
at least on the technical level, ignoring social problems like real-world scammers

edit:
nixos maintainers also censored comments on the [ban milahu commit](https://github.com/NixOS/moderation/commit/836964980e170fe6c856eb0296c27b8337309be8)
see also: [can no longer unhide hidden comments on a commit page](https://github.com/orgs/community/discussions/175529)

**Comment by Kreyren** [comment]
sounds like the thing i would get banned for :D

Make that a merge request to an upstream branch that is synced with upstream and ping us

**Comment by milahu** [comment]
nah, please help yourself

```
git clone --depth=1 https://github.com/milahu/nur-packages

cp -r nur-packages/pkgs/applications/blockchains/haveno \
  ~/src/nixpkgs/pkgs/applications/blockchains
```

https://github.com/NixOS/moderation/commit/836964980e170fe6c856eb0296c27b8337309be8#commitcomment-144375785

> rationale for milanu ban plz

see my [hate-maintainers-censored](https://github.com/milahu/hate-maintainers-censored)

tldr: im "too rude" in my comments
and someone from "the nix community" stalked some of my social feeds
and he didnt like my opinions

**Comment by Kreyren** [comment]
I don't care for politics in open-source i will work with people who's views i don't agree with and won't obstruct their work, but i won't review if you do the code development in this shitty issue way like it's a kindergarten tossing balls at each other.

I have my merge request submitted you can submit changes to my fork branch for me to merge them and submit them to nixpkgs for example.

**Comment by milahu** [comment]
> i won't review

ok, good luck : )

i had the impression that the people i pinged
have an interest in getting haveno run on nixos

i have spent enough time on haveno already
only to see that it breaks on runtime...

**Comment by Kreyren** [comment]
interest in haveno running **without issues and implemented in a sane and controlled way**

**Comment by Kreyren** [comment]
Why the long and hacky unpack phase? Should be handled via https://github.com/NixOS/nixpkgs/pull/272380

As it's likely a complexity-induced issue that is not obvious to see rn

**Comment by milahu** [comment]
> Why the long and hacky unpack phase

you mean `mkOverride`?
that is a missing feature of `gradle2nix`, see https://github.com/tadfisher/gradle2nix/pull/62#issuecomment-2218670780

i dont remember whether this works with `gradle2nix` v1
my version is based on `gradle2nix` v2 with better dependency resolution

but thats implementation detail
some of the jar files contain libraries and executables (libs and bins)
which require patching with `patchelf`

like i said, before spending more time on this
i would recommend to verify if `haveno-reto` is actually working
for example on [arch linux](https://aur.archlinux.org/packages/haveno-reto)
(but remember, dont trust binary releases, treat it like malware)

**Comment by Kreyren** [comment]
screw gradle2nix you should be using the provided gradle setup hook

**Comment by milahu** [comment]
oh man, im so happy that i "left" nixpkgs...
these discussions are so braindead

its a matter of taste which is better
nixpkgs prefers one derivation for all dependencies (vendor hash)
i prefer one derivation per dependency (lockfiles in nix repo)

but thats implementation detail...
if haveno is broken, "the provided gradle setup hook" wont save you

**Comment by nahuhh** [comment]
> or a problem with the haveno network

correct

wait a few days and try again with new version

**Comment by Kreyren** [comment]
> oh man, im so happy that i "left" nixpkgs... these discussions are so braindead
> 
> its a matter of taste which is better nixpkgs prefers one derivation for all dependencies (vendor hash) i prefer one derivation per dependency (lockfiles in nix repo)
> 
> but thats implementation detail... if haveno is broken, "the provided gradle setup hook" wont save you

it's about consensus on packages if you think that things should be done differently in an objectively better way then there is an RFC for it instead of doing subjective shits just because "i prefer them"

reto works in the present version

**Comment by milahu** [comment]
yeah, i prefer to do my subjective shit
instead of wasting my time debating [matters of taste](https://en.wikipedia.org/wiki/De_gustibus_non_est_disputandum)

but hey, good luck porting reto to nixpkgs

**Comment by boldsuck** [comment]
> ### haveno is not working
> 
> the desktop app is working, but it fails to connect to seednodes i dont know whether this is a problem with the client or a problem with the haveno network
> 
> the gui statusbar says: no seed nodes available
> 
> the seednode seems offline: bugc27z7lqjgpwmhbuu3kuwoq2bhailj573r32jm5ydwrcqrwjtblnid.onion:1002
> 
> ```
> $ torsocks nc bugc27z7lqjgpwmhbuu3kuwoq2bhailj573r32jm5ydwrcqrwjtblnid.onion 1002
> 1720959904 ERROR torsocks[402405]: Host unreachable (in socks5_recv_connect_reply() at socks5.c:539)

Your adjustments up to this point were correct, you should have just left the app running.
The seed nodes are under DoS. Your messages are normal. Just let the app run for 20-30 minutes.

> the main version of haveno is broken by design
It's not broken, it's been running live on the Monero testnet (Stagenet) for 1.5 years and will always be there. It's all documented.

> ### is haveno a scam
> 
> it is possible that haveno is a scam, a waste of time

Haveno-reto is **not** a scam. I have made over 30 trades there.


> its already running on mainnet however not usable because there's no mods.

Mods are in the SimpleX chat or as arbs in the client if there are problems.

https://haveno-reto.com/
[Haveno Reto is LIVE](https://monero.town/post/3143272)

**Comment by milahu** [comment]
any progress with haveno on nixos?

> apparently the seed node is offline

currently i can reach 2 of 8 seed nodes

<details>
<summary>
haveno-ping.sh
</summary>

```
#!/usr/bin/env bash

# https://github.com/retoaccess1/haveno-reto/raw/master/core/src/main/resources/xmr_mainnet.seednodes

seednodes="$(cat <<'EOF'
# nodeaddress.onion:port [(@owner,@backup)]
k6wctqd5l2nhmietzs6zg4pric3ukeg7lazzz67ttbl75qwfr2q4yvqd.onion:2002(@anon1)
4gmfgn22tll7ajw3tdp7nru3fvgh5ukt7w53kfv5ymijldivsqtbzdqd.onion:2003(@anon1)
bugc27z7lqjgpwmhbuu3kuwoq2bhailj573r32jm5ydwrcqrwjtblnid.onion:1002(@anon2)
xephvvzd3orepnny7lbia4nkwie5t7wjivlvvz5lhbsck7ubavystead.onion:9992(@anon3 ,@s0)
g4z6oi2wf62nwztwve6qe2hqswj4ezpom6hn7cuy5cxaidey4us76bid.onion:9993(@anon3 ,@s0)
z47tltuwytd5icqq4hni2ammvlugp6pcwqboeu7ngawruualxjjuu3ad.onion:9992(@anon3 ,@s3)
hxb5h34hjgyraycrrxlz5ar2q77mjgondzicwzayqwwvuaepssrn5zyd.onion:9993(@anon3 ,@s3)
u6wwec5ddxswwyrz7rgzuiwowf33llab57y3xzmwwxvsofq2w4m6ihad.onion:1002(@anon4)
oyui76zng5nhml6xlrysfmyqgoutlhwr3h7p35rlzr5fvtg6tzgavvqd.onion:9992(@anon5)
nri52xbvmga4lm7yg7cwszxz2eamt6bb3pgi6bhkzexgk4flb5e5p7qd.onion:9993(@anon5)
im6hcl7hknvsrsns2newv4orfv3kd2ly5yvqtbfkiyzoohscyp5htgqd.onion:2002(@anon6)
EOF
)"

# netselect is not working with torsocks
#torsocks netselect $(echo "$seednodes" | sed -E 's/\(@anon[0-9]+\)$//')
#torsocks netselect $(echo "$seednodes" | sed -E 's/\(@anon[0-9]+\)$//')

# ping is not working with torsocks
#torsocks ping $(echo "$seednodes" | sed -E 's/\(@anon[0-9]+\)$//')

total=$(echo "$seednodes" | wc -l)

timeout=30

while true; do
  res="$(
    pids=()
    while IFS=: read host port; do
      echo $host $port
      echo asdf | timeout $timeout torsocks nc -v $host $port 2>&1 | grep -v 'ERROR torsocks' &
      pids+=($!)
    done < <(echo "$seednodes" | grep -v '^#' | sed -E 's/\(@anon[0-9]+[^)]*\)$//' | shuf)
    wait ${pids[@]}
  )"
  # Connection to $host (127.xxx.xxx.xxx) $port port [tcp/...] succeeded!
  #good=$(echo "$res" | grep ^Connection | wc -l)
  good=$(echo "$res" | grep -E '^Connection .* succeeded!$' | wc -l)
  # nc: connect to $host (127.xxx.xxx.xxx) port $port (tcp) failed: No route to host
  #bad=$(echo "$res" | grep ^nc | wc -l)
  echo $(date +%Y-%m-%dT%H:%M:%S%z) reachable seednodes: $good of $total
done
```

</details>


```
$ ./haveno-ping.sh 
2024-07-30T20:47:21+0200 reachable seednodes: 0 of 8
2024-07-30T20:47:51+0200 reachable seednodes: 0 of 8
2024-07-30T20:48:22+0200 reachable seednodes: 0 of 8
2024-07-30T20:48:52+0200 reachable seednodes: 1 of 8
2024-07-30T20:49:23+0200 reachable seednodes: 1 of 8
2024-07-30T20:49:53+0200 reachable seednodes: 2 of 8
2024-07-30T20:50:24+0200 reachable seednodes: 2 of 8
2024-07-30T20:50:54+0200 reachable seednodes: 1 of 8
```

**Comment by boldsuck** [comment]
> any progress with haveno on nixos?

No, why I think you are working on it.
Haveno is a Java package that runs on all major OS. IDK why NIXOS can't do that.

> > apparently the seed node is offline

Some seeds are gone or under DOS, but new ones are coming with v1.0.10, because everyone has to generate new .onion addresses for externalTor.

> currently i can reach 2 of 8 seed nodes

The client won't connect to more than 2. After bootstrap you no longer need seed nodes. The network is P2P.

**Comment by milahu** [comment]
> I think you are working on it.

no. i stopped working on `haveno.nix`
because i found nanswap and changenow
which "just work" for small amounts
so i made this issue to share my progress

> IDK why NIXOS can't do that.

because nixos has no FHS paths like `/lib/ld-linux.so.2`
so precompiled bins and libs must be patched with [patchelf](https://github.com/NixOS/patchelf)
see `mkOverride` in my [haveno.nix](https://github.com/milahu/nur-packages/raw/master/pkgs/applications/blockchains/haveno/default.nix)

> precompiled bins and libs

but generally, using precompiled files is a security risk
so ideally, more jar files should be built from source
or constructed from other nix packages (tor, monero, ...)

my `haveno.nix` kind-of works
but 2 weeks ago, it failed to connect to seed nodes
maybe i should have had more patience
maybe its a silent failure in haveno
somebody will find out... ; )

**Comment by milahu** [comment]
[updated](https://github.com/milahu/nur-packages/commit/94a031880adec3e25a211b2d3fd8fb561516c495) to haveno 1.1.0

still get `no seed nodes available`

```
Apr-22 10:02:43.129 [JavaFX Application Thread] INFO  h.n.p2p.peers.getdata.RequestDataHandler: 

>> We send a PreliminaryGetDataRequest to peer placeholder.onion:8000
 
Apr-22 10:02:43.210 [JavaFX Application Thread] INFO  h.n.p2p.peers.getdata.RequestDataHandler: Sending getDataRequest to placeholder.onion:8000 failed. That is expected if the peer is offline.
        getDataRequest=PreliminaryGetDataRequest(supportedCapabilities=[0, 1, 2, 5, 6, 7, 10, 11, 12, 13, 14, 15, 16]).
        Exception=java.lang.RuntimeException: java.io.IOException: Cannot connect to hidden service 
Apr-22 10:02:43.224 [JavaFX Application Thread] WARN  haveno.core.app.P2PNetworkSetup: onNoSeedNodeAvailable 
Apr-22 10:02:43.229 [JavaFX Application Thread] INFO  haveno.core.app.HavenoSetup: walletInitialized=false, p2pNetWorkReady=true 
```

related

- https://github.com/retoaccess1/haveno-reto/issues/8
  - If you can't see many offers after boot, you will slowly receive them from peers with time. This can take up to 30-45 minutes.

ok, lets wait some hours...

```
$ ./haveno-ping.sh
2025-04-22T11:01:21+0200 reachable seednodes: 6 of 12
2025-04-22T11:01:51+0200 reachable seednodes: 6 of 12
2025-04-22T11:02:22+0200 reachable seednodes: 6 of 12
2025-04-22T11:02:52+0200 reachable seednodes: 7 of 12
2025-04-22T11:03:22+0200 reachable seednodes: 7 of 12
2025-04-22T11:03:52+0200 reachable seednodes: 7 of 12
```

**Comment by boldsuck** [comment]
You built something broken shit yourself.

>> We send a PreliminaryGetDataRequest to peer placeholder.onion:8000

> Apr-22 10:02:43.210 [JavaFX Application Thread] INFO  h.n.p2p.peers.getdata.RequestDataHandler: Sending getDataRequest to placeholder.onion:8000 failed. That is expected if the peer is offline.


**placeholder.onion:8000**

**Comment by milahu** [comment]
> You built something broken

yep, something is broken in https://github.com/retoaccess1/haveno-reto

> placeholder.onion

yeah i noticed. i thought the onion address was redacted for logging
but i dont see where the `placeholder` string is coming from, its not in the haveno sources

but now (2 hours later) i also see unredacted onion urls

```
INFO: Retrying request to {}->http://lrrgpezvdrbpoqvkavzobmj7dr2otxc5x6wgktrw337bk6mxsvfp5yid.onion:80
Apr-22 12:51:19.824 [JavaFX Application Thread] WARN  h.core.provider.price.PriceFeedService: We did not receive a response from provider http://lrrgpezvdrbpoqvkavzobmj7dr2otxc5x6wgktrw337bk6mxsvfp5yid.onion/. We select the new provider http://elaxlgigphpicy5q7pi5wkz2ko2vgjbq4576vic7febmx4xcxvk6deqd.onion/ and use that for a new request. 
```

the seed nodes are declared in
https://github.com/retoaccess1/haveno-reto/raw/master/core/src/main/resources/xmr_mainnet.seednodes

https://github.com/retoaccess1/haveno-reto/blob/master/.github/workflows/build.yml
does no patching, it just runs `gradle build`

```
      - name: Build with Gradle
        run: ./gradlew build --stacktrace --scan
```

https://github.com/retoaccess1/haveno-reto/blob/master/gradle/wrapper/gradle-wrapper.properties

```
distributionUrl=https\://services.gradle.org/distributions/gradle-8.6-bin.zip
```

i also use gradle 8.6

i dont have the time to debug a complex java app, so im just leaving this here
feel free to ask for help in https://discourse.nixos.org/ (im banned there)

edit:
still no success after 5 hours

```
Apr-22 16:13:09.527 [JavaFX Application Thread] INFO  haveno.network.p2p.network.Statistic: Accumulated network statistics:
Bytes sent: 0;
Number of sent messages/Sent messages: 0 / {};
Number of sent messages per sec: 0.0;
Bytes received: 0
Number of received messages/Received messages: 0 / {};
Number of received messages per sec: 0.0
 
Apr 22, 2025 4:22:15 PM org.apache.http.impl.execchain.RetryExec execute
INFO: I/O exception (java.net.SocketException) caught when processing request to {}->http://2c6y3sqmknakl3fkuwh4tjhxb2q5isr53dnfcqs33vt3y7elujc6tyad.onion:80: SOCKS server general failure
Apr 22, 2025 4:22:15 PM org.apache.http.impl.execchain.RetryExec execute
INFO: Retrying request to {}->http://2c6y3sqmknakl3fkuwh4tjhxb2q5isr53dnfcqs33vt3y7elujc6tyad.onion:80
```

only monero seems to work

> synced with monero mainnet (via tor) at block 3395669

**Comment by milahu** [comment]
> yep, something is broken in https://github.com/retoaccess1/haveno-reto

no, found a problem with my code:
my `mkOverride` failed to patch the rpath of `tor` and `libssl.so.3`

```
$ find ~/.local/share/Haveno -type f -executable -or -name '*.so*' | xargs ldd | grep -e "^/" -e "not found"
/home/user/.local/share/Haveno/xmr_mainnet/tor/tor:
        libevent-2.1.so.7 => not found
        libssl.so.3 => not found
        libcrypto.so.3 => not found
/home/user/.local/share/Haveno/xmr_mainnet/tor/libssl.so.3:
        libcrypto.so.3 => not found

$ patchelf --print-rpath ~/.local/share/Haveno/xmr_mainnet/tor/tor
/nix/store/hdpaa4dyfak99algfwlpdxxhawqymxjf-libz-1.2.8.2025.03.07/lib:/build/source/native/linux/x64/tor.tar.xz

$ patchelf --print-rpath ~/.local/share/Haveno/xmr_mainnet/tor/libssl.so.3
/build/source/native/linux/x64/tor.tar.xz
```

running the `tor` binary directly fails with

```
$ ~/.local/share/Haveno/xmr_mainnet/tor/tor
/home/user/.local/share/Haveno/xmr_mainnet/tor/tor: error while loading shared libraries: libevent-2.1.so.7: cannot open shared object file: No such file or directory
```

fix: `haveno` should [log errors from broken binaries](https://github.com/haveno-dex/haveno/issues/1718)

output of `mkOverride`

```
setting interpreter of native/linux/x64/tor.tar.xz/tor
searching for dependencies of native/linux/x64/tor.tar.xz/tor
    libz.so.1 -> found: /nix/store/hdpaa4dyfak99algfwlpdxxhawqymxjf-libz-1.2.8.2025.03.07/lib
    libevent-2.1.so.7 -> found: /build/source/native/linux/x64/tor.tar.xz
    libssl.so.3 -> found: /build/source/native/linux/x64/tor.tar.xz
    libcrypto.so.3 -> found: /build/source/native/linux/x64/tor.tar.xz
setting RPATH to: /nix/store/hdpaa4dyfak99algfwlpdxxhawqymxjf-libz-1.2.8.2025.03.07/lib:/build/source/native/linux/x64/tor.tar.xz
searching for dependencies of native/linux/x64/tor.tar.xz/libssl.so.3
    libcrypto.so.3 -> found: /build/source/native/linux/x64/tor.tar.xz
```

https://stackoverflow.com/questions/72374384/patchelf-set-interpreter-to-a-path-relative-to-executable

https://github.com/NixOS/patchelf/pull/118

so the rpath `/build/source/native/linux/x64/tor.tar.xz` should be `$ORIGIN`
so `libevent-2.1.so.7` etc are loaded from the directory of the `tor` binary
= so the `tor` binary is relocatable

```
$ cd ~/.local/share/Haveno/xmr_mainnet/tor

$ patchelf --print-rpath tor
/nix/store/hdpaa4dyfak99algfwlpdxxhawqymxjf-libz-1.2.8.2025.03.07/lib:/build/source/native/linux/x64/tor.tar.xz

$ patchelf --set-rpath '/nix/store/hdpaa4dyfak99algfwlpdxxhawqymxjf-libz-1.2.8.2025.03.07/lib:$ORIGIN' tor

$ ./tor
Apr 22 18:26:51.287 [notice] Tor 0.4.8.14 (git-5d040a975df7a060) running on Linux with Libevent 2.1.12-stable, OpenSSL 3.0.16, Zlib 1.2.8.f-sortix-libz, Liblzma N/A, Libzstd N/A and Glibc 2.40 as libc.

$ patchelf --print-rpath libssl.so.3
/build/source/native/linux/x64/tor.tar.xz

$ ldd libssl.so.3 | grep "not found"
        libcrypto.so.3 => not found

$ patchelf --print-rpath libssl.so.3
/build/source/native/linux/x64/tor.tar.xz

$ patchelf --set-rpath '$ORIGIN' libssl.so.3

$ ldd libssl.so.3 | grep "not found"
```

todo:
use https://github.com/NixOS/patchelf/pull/118 in `mkOverride`
with `patchelf --relative-to-file $file`

**Comment by milahu** [comment]
> todo:
> use [NixOS/patchelf#118](https://github.com/NixOS/patchelf/pull/118) in `mkOverride`
> with `patchelf --relative-to-file $file`

the result is not what i wanted
because also `/nix/store` paths are converted to relative paths
only paths under `/build/source` should be made relative to `'$ORIGIN'`

i ended up fixing rpaths in bash
https://github.com/milahu/nur-packages/commit/721ebb6776c8a508c4ad843a18eb74fc2bee42e4

this should be upstreamed to nixpkgs to fix `autoPatchelf .`
which currently uses absolute paths for bundled libraries
but `/build/source` is not available on runtime

> placeholder.onion:8000

```
Apr-23 12:10:55.895 [JavaFX Application Thread] INFO  h.c.n.p2p.seed.DefaultSeedNodeRepository: Seed nodes: [placeholder.onion:8000] 
```

[DefaultSeedNodeRepository.java](https://github.com/haveno-dex/haveno/blob/de5250e89a613d19cc3fd1a34c3168fcef3f619e/core/src/main/java/haveno/core/network/p2p/seed/DefaultSeedNodeRepository.java#L74)

> the seed nodes are declared in
> https://github.com/retoaccess1/haveno-reto/raw/master/core/src/main/resources/xmr_mainnet.seednodes

it works when i pass seed nodes manually

> `haveno-desktop --seedNodes="$(curl -sL https://github.com/retoaccess1/haveno-reto/raw/master/core/src/main/resources/xmr_mainnet.seednodes | grep -v '^#' | cut -d'(' -f1 | tr $'\n' ,)"`

but that `xmr_mainnet.seednodes` file does not end up in my `core.jar`
instead i get [xmr_mainnet.seednodes](https://github.com/haveno-dex/haveno/blob/master/core/src/main/resources/xmr_mainnet.seednodes) from the `haveno-dex/haveno` repo

```
$ unzip -q -c result/opt/haveno/lib/core.jar xmr_mainnet.seednodes 
# nodeaddress.onion:port [(@owner,@backup)]
placeholder.onion:8000 (@placeholder)
```

fix: use [git tag](https://github.com/retoaccess1/haveno-reto/tags) `1.1.0` not `v1.1.0`
https://github.com/milahu/nur-packages/commit/b9c9c67d5149d6ffc514d6c61b88e35e8324fdcb

yay! its finally working

so now this should work with [NUR](https://github.com/nix-community/NUR)

```
nix-shell -p nur.repos.milahu.haveno
haveno-desktop
```

**Comment by milahu** [comment]
https://github.com/NixOS/nixpkgs/issues/293842#issuecomment-2276883763

> I wouldn't trust the milahu's implementation for reasons explained.

there is nothing to "trust" here

> I am working on packaging it, but currently running into issues.

how's it going?
last comment 2024-08-09...

i guess as a "temporary" workaround
youre just "trusting" someone else's binary release

https://discourse.nixos.org/t/discourse-via-email/29/14

> Wow those 480 lines… make my eyes bleed

idiot. do you know how many SLOC all the nixpkgs functions have?

> > I wouldn’t trust the milahu’s implementation for reasons explained.
>
> `@KRYREN` I don’t think you did, but I can imagine it has something to do with the difference between [this](https://github.com/NixOS/nixpkgs/issues/50105#issuecomment-1909437434) and [this](https://github.com/milahu/nixpkgs/issues/68#issuecomment-2236240078).

oh you mean the difference
between your defensive gay shit ...

> I'm afraid this is not the form of governance I signed on to, folks -- so this is farewell.
> 
> It has been a great joy and privilege working with all of you. You are elite without being elitist, which is a rare feat. What you've accomplished here is clearly the heir apparent to the Unix philosophy and tradition. The talent density I've encountered here rivals that of Soda Hall and the earliest years of bitcoin-dev. You truly are the best.

... and my aggressive real shit ...

> "they" xD
> i was banned because i give a shit about "modern feelings", i only care about code

well, youre just another useful idiot
who is intolerant of some [personality types](https://milahu.github.io/alchi/src/alchi-types/alchi-types.html)
pushing for more [mental incest](https://www.urbandictionary.com/define.php?term=mental%20incest&defid=18569716)

but nothing new here...

> One very notable pathology is a form of argument that, reduced to essence, runs like this: “Your refusal to acknowledge that you are guilty of {sin,racism,sexism, homophobia,oppression…} confirms that you are guilty of {sin,racism,sexism, homophobia,oppression…}.” I’ve been presented with enough instances of this recently that I’ve decided that it needs a name. I call this general style of argument “kafkatrapping”, and the above the Model A kafkatrap. -- [Eric Steven Raymond](http://esr.ibiblio.org/?p=2122)
